yoyodyne/models/hard_attention.py

```diff
@@ -0,0 +1,481 @@
+        return loss
+
+    def validation_step(self, batch: data.PaddedBatch, batch_idx: int) -> Dict:
+        predictions, likelihood = self(batch)
+        # Processes for accuracy calculation.
+        val_eval_items_dict = {}
+        for evaluator in self.evaluators:
+            predictions = evaluator.finalize_predictions(
+                predictions, self.end_idx, self.pad_idx
+            )
+            golds = evaluator.finalize_golds(
+                batch.target.padded, self.end_idx, self.pad_idx
+            )
+            val_eval_items_dict[evaluator.name] = evaluator.get_eval_item(
+                predictions, golds, self.pad_idx
+            )
+        val_eval_items_dict.update({"val_loss": -likelihood.mean()})
+        return val_eval_items_dict
+
+    def predict_step(self, batch: data.PaddedBatch, batch_idx: int) -> Dict:
+        predictions, _ = self(batch)
+        return predictions
+
+    def get_decoder(self) -> modules.lstm.HardAttentionLSTMDecoder:
+        if self.attention_context > 0:
+            return modules.lstm.ContextHardAttentionLSTMDecoder(
+                pad_idx=self.pad_idx,
+                start_idx=self.start_idx,
+                end_idx=self.end_idx,
+                attention_context=self.attention_context,
+                decoder_input_size=(
+                    self.source_encoder.output_size
+                    + self.features_encoder.output_size
+                    if self.has_features_encoder
+                    else self.source_encoder.output_size
+                ),
+                num_embeddings=self.target_vocab_size,
+                dropout=self.dropout,
+                bidirectional=False,
+                embedding_size=self.embedding_size,
+                layers=self.decoder_layers,
+                hidden_size=self.hidden_size,
+            )
+        else:
+            return modules.lstm.HardAttentionLSTMDecoder(
+                pad_idx=self.pad_idx,
+                start_idx=self.start_idx,
+                end_idx=self.end_idx,
+                decoder_input_size=(
+                    self.source_encoder.output_size
+                    + self.features_encoder.output_size
+                    if self.has_features_encoder
+                    else self.source_encoder.output_size
+                ),
+                num_embeddings=self.target_vocab_size,
+                dropout=self.dropout,
+                bidirectional=False,
+                embedding_size=self.embedding_size,
+                layers=self.decoder_layers,
+                hidden_size=self.hidden_size,
+            )
+
+    def _get_loss_func(
+        self,
+    ) -> Callable[[torch.Tensor, torch.Tensor, torch.Tensor], torch.Tensor]:
+        """Returns the actual function used to compute loss.
+
+        Returns:
+                Callable[[torch.Tensor, torch.Tensor], torch.Tensor]:
+                    configured loss function.
+        """
+        return self.loss
+
+    def loss(
```

Comment:
I think `_loss` might be better here. 1) would we ever want to call it publicly? And 2) self.loss sounds like a property.

Reply:
that's a better call.